### ISSUE TITLE
make build output fully compatible with native Node.js ES Module format.

### ISSUE THREAD

**Issue by trusktr** [lume/lume#219]
We need to:

- update to latest `solid-js`, which is now compatible with Node ESM.
- make sure all build files have `.js` extensions in their import specifiers. See https://github.com/jsenv/jsenv-node-module-import-map/issues/20 for some ideas and links.
- undo the `"./": "./dist/"` mapping in the `exports` field of `package.json`, because TypeScript does not yet support `exports` and hence types will not map correctly. For now, consumer code will need to include `dist/` within import specifiers.
  - Tracking issue for this, for later: https://github.com/microsoft/TypeScript/issues/33079

Related: #217 (that depends on native ESM, but in the browser)

**Comment by thescientist13** [comment]
Hey this is Owen from Greenwood 👋 

Just wanted to share that support for export maps in NodeJS <> TypeScript is really good now, as of [TypeScript 4.7](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#package.json-exports-imports-and-self-referencing), including the `types` field, and would align really with this project as it is set to `"type":"module"` already/

So taking this project as an example, the equivalent would look like this.  You might still want the top level `types` field for the [TS badge on NPM](https://github.blog/changelog/2020-12-16-npm-displays-packages-with-bundled-typescript-declarations/).
```json
{
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js"
    },
    "./dist/*": "./dist/*"
  },
}
```

----


As far as ESM alignment, I am not sure how `main` is anymore spec compliant, as in general ESM doesn't allow bare specifiers in the first place, and all of this _package.json_ stuff is all based on NodeJS behaviors / assumptions anyway, but maybe I'm missing something there.  I've seen / encountered lots of projects that are using `exports` quite well, with TypeScript, and still interop well (if not better) by supporting an exports map.

Anyway, just wanted to share some of my thoughts / recent experience in this area if it's useful to you.  ✌

**Comment by trusktr** [comment]
Hello Owen, thanks for chiming in!

I knew TS finally got support for `exports`, which is nice. I still like to avoid it because for hand-written import maps it is one more thing people have to manage, and is inconvenient when, for example, a package has a bunch of entries in the exports fields as opposed to only a single entry (a single entry being a new alternative to the `main` field plus hiding all other files). 

The `main` field isn't part of ES spec per se, but for people who are writing vanilla import maps manually, it's a nice go-to field for determining what file to point a bare specifier to in an importmap, and for people not using exports who aren't trying to hide files (which is moot with importmaps in a browser) it's also a little easier (node and build tools fall back to `main` in the absence of `exports`).

I've also still encountered other build tools at organizations just recently that break in the presence of exports fields. It's been an ongoing pain to support libs that choose to adopt Node.js features when, ironically, those libs were meant for browsers. There's just hardly anything to gain for browser libs using exports fields (file hiding is not a feature of importmap spec).

Lume is a browser lib so `exports` don't help much with anything.

---

Even more annoyingly, some "modern" tools break when generating import maps for browsers because they unconditionally require Node's `exports` as a standard, making people have to fall back to manual handling of import maps which defeats the purpose. F.e. this breaks:

https://generator.jspm.io/#Y2VhYGBkDM0rySzJSU1hcMgpzU3Vz8ksS9VNzk9JdTDQM9MzBwDCNxWRJAA

Here's an example that doesn't have the issue:

https://generator.jspm.io/#U2VhYGBkDM0rySzJSU1hyCnNTXUw0DPWM9BNzCnISNQzMQEADGFWCCIA

With `@lume/live-code` I want to avoid importing the `dist/index.js` file specified in the dependency's `exports` field here,

https://ga.jspm.io/npm:thememirror@2.0.1/package.json

as that imports a bunch of themes I don't need. In that case I can't use JSPM Generator, but am forced to write a map manually (or find some other tool).

TLDR its simple to just use `main` with `types` because it aligns with the simple concept that with manual vanilla importmaps we have to write an entry for each bare specifier of a package anyway.

Typically I have one or both of the following importmap lines for any of my libs depending on if they have an index, separate files to import from inside (wouldn't want to write those all in `exports)`, or both:

```
"some-lib": "./node_modules/some-lib/dist/index.js",
"some-lib/": "./node_modules/some-lib/"
```

and that ends up being an easy pattern to follow across the board for simplicity.

**Comment by trusktr** [comment]
Btw, we've been slowly chipping away at SSR for Lume in Astro, and eventually we can try to re-use some of the logic in Greenwood. That'll be sweet.

**Comment by thescientist13** [comment]
Hey @trusktr , just wanted to give you an update.  I've got the Hello World demo (mostly*) working with Greenwood now!
- https://github.com/thescientist13/greenwood-lume
- https://greenwood-lume.vercel.app/

I was able to [enhance some things](https://github.com/ProjectEvergreen/greenwood/discussions/1396) in Greenwood's import map generation logic, including a feature to "auto expand" `main` entry points to create import map entries for all files found within that entry.  Should add a little more compatibility to CJS packages in general too. 🤞 

Cool project, and had a lot of fun getting this all put together and getting it working with import maps, and hopefully adds broader ecosystem compatibility over all.

> Did open one issue to the docs repo that may be of interest based on my experience - https://github.com/lume/lume.github.io/issues/3

----

*regarding mostly working, there are no errors and it's definitely doing _something_, but it doesn't behave exactly like the demo you have.

Not sure if you have any thoughts?  I got a little write up here with what I am seeing if you have any thoughts
https://github.com/thescientist13/greenwood-lume/issues/2

Gonna try and hack on it a little more tonight

**Comment by trusktr** [comment]
@thescientist13 oh hey, that's awesome! I need to work on making notifications for GitHub work better, so I don't miss things!

I see the earth demo expands, which is due to the lume-scene by default taking up 100% size of its parent out of the box using JavaScript, so it reads the parent size, sets its own size, this causes parent to grow (body grows vertically by default based on children), causing an infinite growing loop.

I will change this so the lume-scene is sized with CSS like with canvas, while still allowing its sizing feature to work when nested in other Lume 3D nodes.

I will circle back to this to check it out more closely!

**Comment by thescientist13** [comment]
Ah, can't believe it was that simple!  Thanks, I got it looking nice now ✨ 
https://greenwood-lume.vercel.app/

Also, if you're interested, I made plugin for Astro <> WCC integration (WCC is the primary SSR lib under the hood of Greenwood).  Nothing super fancy yet, but can do basic SSR at least
https://github.com/ProjectEvergreen/astro-wcc
